src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/Grpc/GrpcStatusDeserializer.cs

```diff
@@ -0,0 +1,292 @@
+            }
+        }
+
+        return duration;
+    }
+
+    private static Any DecodeAny(Stream stream)
+    {
+        var length = DecodeVarint(stream);
+        var endPosition = stream.Position + length;
+        Any any = default;
+
+        while (stream.Position < endPosition)
+        {
+            var tag = DecodeTag(stream);
+            var fieldNumber = tag >> 3;
+            var wireType = tag & 0x7;
+
+            switch (fieldNumber)
+            {
+                case 1: // type_url
+                    any.TypeUrl = DecodeString(stream);
+                    break;
+                case 2: // value
+                    any.Value = DecodeBytes(stream);
+                    break;
+                default:
+                    SkipField(stream, wireType);
+                    break;
+            }
+        }
+
+        return any;
+    }
+
+    private static uint DecodeTag(Stream stream)
+    {
+        return (uint)DecodeVarint(stream);
+    }
+
+    private static long DecodeVarint(Stream stream)
+    {
+        long result = 0;
+        int shift = 0;
+
+        while (true)
+        {
+            int b = stream.ReadByte();
+            if (b == -1)
+            {
+                throw new EndOfStreamException();
+            }
+
+            result |= (long)(b & 0x7F) << shift;
+            if ((b & 0x80) == 0)
+            {
+                return result;
+            }
+
+            shift += 7;
+            if (shift >= 64)
+            {
+                throw new InvalidDataException("Invalid varint");
+            }
+        }
+    }
+
+    private static int DecodeInt32(Stream stream) => (int)DecodeVarint(stream);
+
+    private static long DecodeInt64(Stream stream) => DecodeVarint(stream);
+
+    private static string DecodeString(Stream stream)
+    {
+        var bytes = DecodeBytes(stream);
+        return Encoding.UTF8.GetString(bytes);
+    }
+
+    private static byte[] DecodeBytes(Stream stream)
+    {
+        var length = (int)DecodeVarint(stream);
+        var buffer = new byte[length];
+        int read = stream.Read(buffer, 0, length);
+        if (read != length)
+        {
+            throw new EndOfStreamException();
+        }
+
+        return buffer;
+    }
+
+    private static void SkipField(Stream stream, uint wireType)
+    {
+        switch (wireType)
+        {
+            case WIRETYPE_VARINT:
+                DecodeVarint(stream);
+                break;
+            case WIRETYPE_FIXED64:
+                stream.Position += 8;
+                break;
+            case WIRETYPE_LENGTH_DELIMITED:
+                var length = DecodeVarint(stream);
+                stream.Position += length;
+                break;
+            case WIRETYPE_FIXED32:
+                stream.Position += 4;
+                break;
+            default:
+                throw new InvalidDataException($"Unknown wire type: {wireType}");
+        }
+    }
+
+    internal struct Duration
```

Comment:
Can we declare these `internal readonly struct`?

Reply:
Added `readonly` to all the `struct` expect status as it contains mutable List in it.